nabs/simulators.py

```diff
@@ -0,0 +1,244 @@
+            # of how they were staged.
+            staged_devices.pop()
+
+    # at end of plan, nothing should be left
+    if open_stack:
+        raise ValidError("Plan ended without all runs being closed.")
+    if run_keys:
+        raise ValidError("Plan ended with unmatched run keys.")
+    if staged_devices:
+        raise ValidError(
+            "Plan ended without unstaging all staged devices."
+        )
+
+
+def check_stray_calls(
+    plan: Iterator[Any],
+    patches: List[str] = default_patches
+) -> None:
+    """
+    Validate that plan does not invoke any caput functionality
+    outside of messages.
+
+    This is rather jank currently, it's entirely possible there is a
+    better way around this.
+
+    Relies on the pre-existing knowledge of which methods make calls
+    to pyepics/caput functionality.
+
+    Parameters
+    ----------
+    plan : iterable or generator
+        Must yield `Msg` objects
+
+    Raises
+    ------
+    ValidError
+        If attempts to access any forbidden methods
+    """
+
+    with patch_sys_modules(patches):
+        for _ in plan:
+            continue
+
+
+# check_limits is not hinted, so hinting this becomes miserable
+validators = [
+    check_stray_calls,
+    check_open_close,
+    check_limits,
+]
+
+
+def validate_plan(
+    plan: Generator,
+    validators=validators
+) -> Tuple[bool, str]:
+    """
+    Validate plan with all available checkers.
+
+    Parameters
+    ----------
+    plan: generator function
+        Once called, must yield `Msg` objects.
+
+    validators: list of check functions
+        functions to run on the provided plan.  These should take an
+        evaluated plan (generator) as input, and raise exceptions on
+        failure.  (should NOT take generator functions)
+
+    Returns
+    -------
+    boolean
+        Indicates if validation was successful (``True``) or failed
+        (``False``).
+    str
+        Error message that explains the reason for validation
+        failure. Empty string if validation is successful.
+
+    """
+    success, msg = True, ""
+    try:
+        plan_list = itertools.tee(plan, len(validators))
+        for i, check in enumerate(validators):
+            print(f'running {check.__name__}')
+            check(plan_list[i])
+    except Exception as ex:
+        print(ex)
+        msg = (f'Plan validation failed for reason: {str(ex)}')
+        success = False
+
+    return success, msg
+
+
+def summarize_plan(plan: Generator):
+    """Print summary of plan
+
+    Prints a minimal version of the plan, showing only moves and
+    where events are created.
+
+    Taken from bluesky.simulators.summarize_plan, but adapted to
+    print lcls-daq information specifically
+
+    Parameters
+    ----------
+    plan : iterable
+        Must yield `Msg` objects
+    """
+    read_cache: List[str] = []
+    daq_keys = ['events', 'record', 'use_l3t', 'duration']
+    daq_cfg = {k: None for k in daq_keys}
```

Comment:
This misses any stateful nature of the DAQ set up prior to the plan, but I think that's probably OK. Including this information is a bit gross and requires finding the registered daq instance and populating the starting config differently here.

Reply:
My thought on this was that we're only aiming to summarize what the plan itself does.  If it doesn't configure the DAQ we shouldn't report its configuration and potentially mislead someone into thinking it does.